dwave/optimization/src/nodes/manipulation.cpp

```diff
@@ -442,8 +486,34 @@ void PutNode::propagate(State& state) const {
 
 void PutNode::revert(State& state) const { return data_ptr<PutNodeState>(state)->revert(); }
 
-ReshapeNode::ReshapeNode(ArrayNode* node_ptr, std::span<const ssize_t> shape)
-        : ArrayOutputMixin(shape), array_ptr_(node_ptr) {
+// Reshape allows one shape dimension to be -1. In that case the size is inferred.
+// We do that inference here.
+std::vector<ssize_t> infer_reshape(Array* array_ptr, std::vector<ssize_t>&& shape) {
+    // if the base array is dynamic, we might allow the first dimension to be negative
+    // 1. So let's defer to the various constructors.
+    if (array_ptr->dynamic()) return shape;
+
+    // Check if there are any -1s, and if not fallback to other input checking.
+    auto it = std::ranges::find(shape, -1);
+    if (it == shape.end()) return shape;
+
```

Comment:
```suggestion

// Ensure there is at most one -1
if (++it != shape.end()) {
    // Same error message as NumPy
    throw std::invalid_argument("can only specify one unknown dimension");
}
```

Reply:
This actually gets checked by the `ArrayOutputMixin` because that doesn't allow any `-1`. So I am deferring the error to there. The error message is slightly less explicit, but IMO it's OK.

Reply:
I think it checks that it's not dynamic, i.e. the first index is not negative, but that won't catch the case of multiple -1s...

Reply:
Actually, I don't think it gets checked at all. Just returns `DYNAMIC_SIZE` if the first index is negative. So then the `this->dynamic()` check below would return true and throw an error if there are an even number of -1s in the shape, but not if there are an odd number

Reply:
Let me add a test case

Reply:
4442ba1 adds a test case, it does correctly throw.

Reply:
It's because it results in a negative or otherwise non-matching shape. So it fails on the check for consistency between `Reshape`'s shape and its predecessor's shape.

Reply:
Ok, I found an edge case, `ReshapeNode(&A, {12, -1, -1}`. Will fix.

Reply:
Fixed by 6439503